src/Razor/src/Microsoft.VisualStudio.Editor.Razor/DefaultDocumentDivergenceChecker.cs

```diff
@@ -0,0 +1,304 @@
+                var root = parsedText.GetRoot();
+                var childNodes = root.ChildNodes();
+                var properties = childNodes.Where(node => node.Kind() == CodeAnalysis.CSharp.SyntaxKind.PropertyDeclaration).OfType<PropertyDeclarationSyntax>();
+                propertyList.AddRange(properties);
+            }
+
+            return propertyList;
+        }
+
+        private bool TryGetSyntaxTreeAndSource(DocumentSnapshot document, out RazorSyntaxTree syntaxTree, out SourceText sourceText)
+        {
+            if (document.TryGetGeneratedOutput(out var codeDocument) &&
+                document.TryGetText(out sourceText))
+            {
+                syntaxTree = codeDocument.GetSyntaxTree();
+                return true;
+            }
+            else
+            {
+                if (!document.TryGetText(out sourceText))
+                {
+                    // Can't get the source text synchronously
+                    syntaxTree = null;
+                    return false;
+                }
+
+                syntaxTree = ParseSourceText(document.FilePath, sourceText);
+                return true;
+            }
+        }
+
+        private RazorSyntaxTree ParseSourceText(string filePath, SourceText sourceText)
+        {
+            var sourceDocument = new SourceTextSourceDocument(filePath, sourceText);
+            var syntaxTree = RazorSyntaxTree.Parse(sourceDocument, ParserOptions);
+            return syntaxTree;
+        }
+
+        private class SourceTextSourceDocument : RazorSourceDocument
+        {
+            private readonly string _filePath;
+            private readonly SourceText _sourceText;
+            private readonly RazorSourceTextLineCollection _lines;
+
+            public SourceTextSourceDocument(string filePath, SourceText sourceText)
+            {
+                if (filePath is null)
+                {
+                    throw new ArgumentNullException(nameof(filePath));
+                }
+
+                if (sourceText is null)
+                {
+                    throw new ArgumentNullException(nameof(sourceText));
+                }
+
+                _filePath = filePath;
+                _sourceText = sourceText;
+                _lines = new RazorSourceTextLineCollection(_sourceText.Lines);
+            }
+
+            public override char this[int position] => _sourceText[position];
+
+            public override Encoding Encoding => _sourceText.Encoding;
+
+            public override string FilePath => _filePath;
+
+            public override int Length => _sourceText.Length;
+
+            public override RazorSourceLineCollection Lines => _lines;
+
+            public override void CopyTo(int sourceIndex, char[] destination, int destinationIndex, int count) => _sourceText.CopyTo(sourceIndex, destination, destinationIndex, count);
+
+            public override byte[] GetChecksum() => Array.Empty<byte>();
```

Comment:
I don't know what specifically it would break, but this seems dangerous to me. Can we make it pass through to `_sourceText` or something?

Reply:
Nah not dangerous, wedon't want it to calculate the checksum anyways because checksums are only used for debugging purposes. We're just trying to parse the world to introspect on some details not enable debugging.